src/BizHawk.Client.Common/movie/tasproj/ZwinderStateManager.cs

```diff
@@ -311,6 +307,56 @@ public void EvictReserved(int frame)
 			}
 		}
 
+		public void Unreserve(int frame)
+		{
+			// Before removing the state, check if it should still be reserved.
+			// For now, this just means checking if we need to keep this state to satisfy the ancient interval.
+			if (ShouldKeepForAncient(frame))
+				return;
+
+			EvictReserved(frame);
+		}
+
+		/// <summary>
+		/// This method is only to be used as the capture callback for a ZwinderBuffer, since it assumes the state is about to be removed from the buffer.
+		/// </summary>
+		private void HandleStateDecay(ZwinderBuffer buffer, int stateIndex)
+		{
+			var state = buffer.GetState(stateIndex);
+
+			// Add to reserved buffer if externally reserved, or if it matches an "ancient" state consideration
+			if (_reserveCallback(state.Frame) || ShouldKeepForAncient(state.Frame))
+				AddToReserved(state);
+			else
+				// We remove from the state cache because this state is about to be removed, by the buffer.
+				StateCache.Remove(state.Frame);
+		}
+
+		/// <summary>
+		/// Will removing the state on this leave us with a gap larger than the ancient interval?
+		/// </summary>
+		private bool ShouldKeepForAncient(int frame)
+		{
+			int index = StateCache.BinarySearch(frame + 1);
+			if (index < 0)
+				index = ~index;
+			if (index == StateCache.Count)
+			{
+				// There is no future state, so there is no gap between states for us to measure.
+				// We're probably unreserving for a marker removal. Allow it to be removed, so we don't pollute _reserved.
+				return true;
+			}
+			if (index <= 1)
+				// index == 0 should not be possible. (It's the index of the state after the given frame.)
+				// index == 1 would mean we are considering removing the state on frame 0.
+				// We must always have a state on frame 0.
+				return true;

```

Comment:
If I understand this method correctly, isn't this branch reachable by `Unreserve(1);` --> `[ 0, 2, 3 ].BinarySearch(1 + 1)`?

Reply:
Technically true, but it doesn't really matter. This implies `Unreserve(1)` is called while there isn't actually a state on frame 1, in which case there is nothing to do.